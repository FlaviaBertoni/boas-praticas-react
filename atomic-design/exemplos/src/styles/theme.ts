const colors = {
  white: '#fff',
  black: '#000',
  primary: '#7E7EB8',
  dark: '#2C2E33',
  disable: '#747A88',
  'card-bg': '#F8F8FA',
  transparent: 'transparent',
};

const getColorCode = (color?: ThemeColorsKeys) => colors[color || 'primary'];

const isDarkColor = (color?: ThemeColorsKeys) =>
  ![colors.white, colors.disable].includes(getColorCode(color) || colors.primary);

const theme = { colors, getColorCode, isDarkColor };

export type ThemeType = typeof theme;
export type ThemeColorsType = typeof colors;
export type ThemeColorsKeys = keyof ThemeColorsType;

export default theme;
