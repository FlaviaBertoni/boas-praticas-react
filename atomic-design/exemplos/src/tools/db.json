{
  "vehicles": [
    {
      "id": "1",
      "model": "K.I.T.T",
      "price": 42,
      "transmission": "Manual",
      "seats": 4,
      "fuel_efficiency": 14.8,
      "efficiency_unit": "km",
      "image": "/images/cars/KITT.png"
    },
    {
      "id": "2",
      "model": "Ecto-1",
      "price": 52.99,
      "transmission": "Manual",
      "seats": 4,
      "fuel_efficiency": 7,
      "efficiency_unit": "km",
      "image": "/images/cars/ecto-1.png"
    },
    {
      "id": "3",
      "model": "DeLorean",
      "price": 52,
      "transmission": "Manual",
      "seats": 2,
      "fuel_efficiency": 12.8,
      "efficiency_unit": "km",
      "image": "/images/cars/delorean.png"
    },
    {
      "id": "4",
      "model": "Bat MÃ³vel",
      "price": 42,
      "transmission": "Auto",
      "seats": 2,
      "fuel_efficiency": 5.8,
      "efficiency_unit": "km",
      "image": "/images/cars/Batmovel.png"
    },
    {
      "id": "5",
      "model": "Herbie",
      "price": 79,
      "transmission": "Manual",
      "seats": 4,
      "fuel_efficiency": 15.6,
      "efficiency_unit": "km",
      "image": "/images/cars/Herbie.png"
    },
    {
      "id": "6",
      "model": "The Mistery Machine",
      "price": 199,
      "transmission": "Auto",
      "seats": 7,
      "fuel_efficiency": 9.5,
      "efficiency_unit": "km",
      "image": "/images/cars/TheMysteryMachine.png"
    }
  ],
  "users": [
    {
      "id": 1,
      "email": "carlostgti@outlook.com",
      "password": "M@n4nc14l",
      "name": "Carlos Queiroz"
    },
    {
      "id": 2,
      "email": "cqueiroz@ciandt.com",
      "password": "12345678",
      "name": "Carlos Queiroz"
    },
    {
      "newUser": {
        "name": "Carlos teste",
        "email": "carlos.da.catia@gmail.com",
        "password": "123456",
        "id": 1615213221105
      },
      "id": 3
    },
    {
      "newUser": {
        "name": "Carlos Queiroz",
        "email": "carlos.da.catia@gmail.com",
        "password": "123456",
        "id": 1615214235093
      },
      "id": 4
    },
    {
      "newUser": {
        "name": "Carlos Queiroz",
        "email": "carlos.da.catia@gmail.com",
        "password": "123456",
        "id": 1615214778055
      },
      "id": 5
    },
    {
      "name": "Carlos Queiroz",
      "email": "carlos.da.catia@gmail.com",
      "password": "123456",
      "id": 1615215407741
    }
  ]
}