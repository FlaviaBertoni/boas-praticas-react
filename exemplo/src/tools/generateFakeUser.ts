const generateFakeUser = () => ({
  name: 'Fake test user',
  email: 'fakeuser@fake.dev',
  password: 'F@k3p4ssw0rd',
});

export default generateFakeUser;
