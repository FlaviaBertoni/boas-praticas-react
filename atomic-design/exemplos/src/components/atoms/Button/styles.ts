import styled, { css } from 'styled-components';
import { ThemeColorsKeys } from '../../../styles/theme';

interface ButtonProps {
  color?: ThemeColorsKeys;
  width?: string;
}

const BaseButton = styled.button<ButtonProps>`
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  align-items: center;
  justify-content: center;

  ${({ width }) =>
    !!width &&
    css`
      width: ${width};
    `};
`;

export const OutlinedButton = styled(BaseButton)<ButtonProps>`
  border: 1px solid ${({ color, theme }) => theme.getColorCode(color)};
  color: ${({ color, theme }) => theme.getColorCode(color)};
`;

export const TextButton = styled(BaseButton)<ButtonProps>`
  border: 1px solid transparent;
  color: ${({ color, theme }) => theme.getColorCode(color)};
`;

export const ContainedButton = styled(BaseButton)<ButtonProps>`
  border: 1px solid transparent;
  color: ${({ color, theme }) => (theme.isDarkColor(color) ? theme.colors.white : theme.colors.black)};
  background-color: ${({ color, theme }) => theme.getColorCode(color)};
`;
