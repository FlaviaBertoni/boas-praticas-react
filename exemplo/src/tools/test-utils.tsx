import { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from '../styles/theme';

const AllTheProviders: FC = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
