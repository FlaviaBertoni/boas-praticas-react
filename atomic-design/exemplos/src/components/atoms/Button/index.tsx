import { ThemeColorsKeys } from '../../../styles/theme';
import { ContainedButton, OutlinedButton, TextButton } from './styles';

export type buttonVariantType = 'text' | 'outlined' | 'contained';
interface ButtonProps {
  color?: ThemeColorsKeys;
  variant?: buttonVariantType;
  width?: string;
}

const Button: React.FC<ButtonProps> = ({ variant, color, width, children }) => {
  const buttonVariant = variant || 'outlined';
  return (
    <>
      {buttonVariant === 'contained' && (
        <ContainedButton color={color} width={width}>
          {children}
        </ContainedButton>
      )}
      {buttonVariant === 'outlined' && (
        <OutlinedButton color={color} width={width}>
          {children}
        </OutlinedButton>
      )}
      {buttonVariant === 'text' && (
        <TextButton color={color} width={width}>
          {children}
        </TextButton>
      )}
    </>
  );
};

export default Button;
