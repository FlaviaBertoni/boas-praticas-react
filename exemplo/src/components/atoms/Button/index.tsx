import { ThemeColorsKeys } from '../../../styles/theme';
import { ContainedButton, OutlinedButton, TextButton } from './styles';

export type ButtonVariants = 'text' | 'outlined' | 'contained';
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  color?: ThemeColorsKeys;
  variant?: ButtonVariants;
  width?: string;
  margin?: string;
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

const Button: React.FC<ButtonProps> = ({ variant, color, width, className, children, margin, onClick, ...rest }) => {
  const buttonVariant = variant || 'outlined';
  return (
    <>
      {buttonVariant === 'contained' && (
        <ContainedButton onClick={onClick} color={color} width={width} margin={margin} className={className} {...rest}>
          {children}
        </ContainedButton>
      )}
      {buttonVariant === 'outlined' && (
        <OutlinedButton onClick={onClick} color={color} width={width} margin={margin} className={className} {...rest}>
          {children}
        </OutlinedButton>
      )}
      {buttonVariant === 'text' && (
        <TextButton onClick={onClick} color={color} width={width} margin={margin} className={className} {...rest}>
          {children}
        </TextButton>
      )}
    </>
  );
};

export default Button;
