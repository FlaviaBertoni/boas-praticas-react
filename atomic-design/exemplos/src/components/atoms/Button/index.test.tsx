import { render } from '../../../tools/test-utils';
import Button, { buttonVariantType } from '.';
import theme, { ThemeColorsType } from '../../../styles/theme';

const colors = Object.keys(theme.colors);
const variants: buttonVariantType[] = ['text', 'outlined', 'contained'];

describe('Test case for Button component', () => {
  it('should be able to render the Button component without width', () => {
    const { container } = render(<Button>Text</Button>);

    // these assertions were using jest-style-component lib
    expect(container.firstChild).not.toHaveStyleRule('width', '150px');
  });

  it('should be able to render the Button component with width', () => {
    const { container } = render(<Button width="150px">Text</Button>);

    // these assertions were using jest-style-component lib
    expect(container.firstChild).toHaveStyleRule('width', '150px');
  });

  // We can test a list of properties using each.
  it.each(colors)('should be able to render the Button component for %s color and match with the snapshot', color => {
    const { container } = render(<Button color={color as keyof ThemeColorsType}>unit test</Button>);

    expect(container).toMatchSnapshot();
  });

  // We can test a list of properties using each.
  it.each(variants)(
    'should be able to render the Button component for %s variant and match with the snapshot',
    variant => {
      const { container } = render(<Button variant={variant}>unit test</Button>);

      expect(container).toMatchSnapshot();
    },
  );
});
