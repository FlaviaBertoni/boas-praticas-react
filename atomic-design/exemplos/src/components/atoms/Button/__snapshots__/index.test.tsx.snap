// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for Button component should be able to render the Button component for black color and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid #000;
  color: #000;
}

<div>
  <button
    class="c0 c1"
    color="black"
  >
    unit test
  </button>
</div>
`;

exports[`Test case for Button component should be able to render the Button component for contained variant and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid transparent;
  color: #fff;
  background-color: #7E7EB8;
}

<div>
  <button
    class="c0 c1"
  >
    unit test
  </button>
</div>
`;

exports[`Test case for Button component should be able to render the Button component for dark color and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid #2C2E33;
  color: #2C2E33;
}

<div>
  <button
    class="c0 c1"
    color="dark"
  >
    unit test
  </button>
</div>
`;

exports[`Test case for Button component should be able to render the Button component for disable color and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid #747A88;
  color: #747A88;
}

<div>
  <button
    class="c0 c1"
    color="disable"
  >
    unit test
  </button>
</div>
`;

exports[`Test case for Button component should be able to render the Button component for outlined variant and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid #7E7EB8;
  color: #7E7EB8;
}

<div>
  <button
    class="c0 c1"
  >
    unit test
  </button>
</div>
`;

exports[`Test case for Button component should be able to render the Button component for primary color and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid #7E7EB8;
  color: #7E7EB8;
}

<div>
  <button
    class="c0 c1"
    color="primary"
  >
    unit test
  </button>
</div>
`;

exports[`Test case for Button component should be able to render the Button component for text variant and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid transparent;
  color: #7E7EB8;
}

<div>
  <button
    class="c0 c1"
  >
    unit test
  </button>
</div>
`;

exports[`Test case for Button component should be able to render the Button component for white color and match with the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  border-radius: 16px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid #fff;
  color: #fff;
}

<div>
  <button
    class="c0 c1"
    color="white"
  >
    unit test
  </button>
</div>
`;
